src/owl/core/owl_tensor_impl.c

```diff
@@ -0,0 +1,549 @@
+    argv[8], argv[9], argv[10], argv[11], argv[12], argv[13], argv[14]
+  );
+}
+
+
+value FUN_NATIVE (spatial_backward) (
+  value vInput, value vOutput_back, value vInput_back,
+  value vBatches, value vInput_cols, value vInput_rows, value vIn_channel,
+  value vKernel_cols, value vKernel_rows,
+  value vOutput_cols, value vOutput_rows,
+  value vRow_stride,  value vCol_stride,
+  value vPad_rows, value vPad_cols
+) {
+  struct caml_ba_array *IN  = Caml_ba_array_val(vInput);
+  struct caml_ba_array *OUB = Caml_ba_array_val(vOutput_back);
+  struct caml_ba_array *INB = Caml_ba_array_val(vInput_back);
+  TYPE *input_ptr = (TYPE *) IN->data;
+  TYPE *output_backward_ptr = (TYPE *) OUB->data;
+  TYPE *input_backward_ptr  = (TYPE *) INB->data;
+
+  int batches       = Long_val(vBatches);
+  int input_cols    = Long_val(vInput_cols);
+  int input_rows    = Long_val(vInput_rows);
+  int in_channel    = Long_val(vIn_channel);
+  int kernel_cols   = Long_val(vKernel_cols);
+  int kernel_rows   = Long_val(vKernel_rows);
+  int output_cols   = Long_val(vOutput_cols);
+  int output_rows   = Long_val(vOutput_rows);
+  int row_stride    = Long_val(vRow_stride);
+  int col_stride    = Long_val(vCol_stride);
+  int pad_rows      = Long_val(vPad_rows);
+  int pad_cols      = Long_val(vPad_cols);
+
+  const int ksize      = kernel_cols * kernel_rows;
+  const int output_cri = output_cols * output_rows * in_channel;
+  const int output_ri  = output_rows * in_channel;
+  const int input_cri  = input_cols  * input_rows  * in_channel;
+  const int input_ri   = input_rows  * in_channel;
+
+  if (pad_cols < 0) pad_cols = 0;
+  if (pad_rows < 0) pad_rows = 0;
+
+  memset(input_backward_ptr, 0,
+    batches * input_cols * input_rows * in_channel * sizeof(TYPE));
+
+  int i, j, k, l;
+  for (i = 0; i < batches; ++i) {
+    const int input_idx_base = i * input_cri;
+    const int output_idx_base_i = i * output_cri;
+    for (j = 0; j < output_cols; ++j) {
+      const int output_idx_base_j = output_idx_base_i + j * output_ri;
+      for (k = 0; k < output_rows; ++k) {
+        const int output_idx_base = output_idx_base_j + k * in_channel;
+
+        const int cstart = j * col_stride - pad_cols;
+        const int rstart = k * row_stride - pad_rows;
+        const int cend   = cstart + kernel_cols;
+        const int rend   = rstart + kernel_rows;
+
+        for (l = 0; l < in_channel; ++l) {
+          TYPE m;
+          int output_idx = output_idx_base + l;
+          m = *(output_backward_ptr + output_idx);
+
+          int idx[ksize];
+          memset(idx, 0, ksize * sizeof(int));
+
+          TYPE acc = INITACC;
+          int max_idx = 0;
+          int c = 0;
+          for (int a = cstart; a < cend; ++a) {
+            for (int b = rstart; b < rend; ++b) {
+              if (a >= 0 && a < input_cols &&
+                  b >= 0 && b < input_rows) {
+                int input_idx =
+                  input_idx_base + a * input_ri + b * in_channel + l;
```

Comment:
In this PR, this is fine. I will find some time in discussing with you next week, this can be potentially optimised further.